### ISSUE TITLE
synthetics-minion, Crashlooping not enough resources.

### ISSUE THREAD

**Issue by davidbIQ** [newrelic/helm-charts#213]
**Bug description**

I get an error of:  The node where job synthetic-healthcheck-2345bc066c4f453cb246b4879fa34c3e was scheduled did not have enough available resources.

Kubernetes version 1.17.11
ubernetes where the problem occurs.

**Which chart?**
    repository: https://helm-charts.newrelic.com/charts
    name: synthetics-minion
    version: "1.0.18"

**What happened?**
If the node it is scheduling it on is close to full but it can fit, it is slotted on to the node and then I believe it does the synthetic health check it runs in into a resource issue:

```
 kubectl describe po newrelic-synthetics-minion-2  -nmonitoring
Name:           newrelic-synthetics-minion-2
Namespace:      monitoring
Priority:       0
Node:           aks-defaultpool-33270332-vmss000015/10.240.0.9
Start Time:     Tue, 24 Nov 2020 17:21:11 -0600
Labels:         app.kubernetes.io/instance=newrelic-synthetics-minion
                app.kubernetes.io/name=synthetics-minion
                controller-revision-hash=newrelic-synthetics-minion-fb78b49b6
                statefulset.kubernetes.io/pod-name=newrelic-synthetics-minion-2
Annotations:    <none>
Status:         Running
IP:             10.244.8.253
Controlled By:  StatefulSet/newrelic-synthetics-minion
Init Containers:
  update-mounted-subpath-permissions:
    Container ID:  docker://8521fb9b517952a3faf3df5a937c6b206c016b6e8f07dff8c67bd6a156116497
    Image:         quay.io/newrelic/synthetics-minion:3.0.28
    Image ID:      docker-pullable://quay.io/newrelic/synthetics-minion@sha256:8035c60d4d282d3c7a2007d17ed515821de3f123f867930e905e5e0c4eca763d
    Port:          <none>
    Host Port:     <none>
    Command:
      update-k8s-mounted-subpath-permissions
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 24 Nov 2020 17:21:46 -0600
      Finished:     Tue, 24 Nov 2020 17:21:46 -0600
    Ready:          True
    Restart Count:  0
    Environment:
      MINION_UID:                   2379
      MINION_GID:                   3729
      TMP_PATH:                     /tmp
      PERMANENT_DATA_STORAGE_PATH:  /var/lib/newrelic/synthetics
      CUSTOM_MODULES_PATH:          /var/lib/newrelic/synthetics/modules
      USER_DEFINED_VARIABLES_PATH:  /var/lib/newrelic/synthetics/variables
    Mounts:
      /tmp from minion-volume (rw,path="newrelic-synthetics-minion/tmp")
      /var/run/secrets/kubernetes.io/serviceaccount from newrelic-synthetics-minion-token-xppxf (ro)
Containers:
  synthetics-minion:
    Container ID:  docker://02c931329f4199fd104b39000cadbd17809a00d1ab95ec5c3180d61792306c97
    Image:         quay.io/newrelic/synthetics-minion:3.0.28
    Image ID:      docker-pullable://quay.io/newrelic/synthetics-minion@sha256:8035c60d4d282d3c7a2007d17ed515821de3f123f867930e905e5e0c4eca763d
    Ports:         8080/TCP, 8180/TCP, 5005/TCP, 65100/TCP, 65101/TCP, 65102/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP, 0/TCP, 0/TCP, 0/TCP
    State:         Terminated
      Reason:      Error
      Message:     .fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:568)
! at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:507)
! at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.doGetLog(PodOperationsImpl.java:186)
! at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.getLog(PodOperationsImpl.java:195)
! at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.getLog(PodOperationsImpl.java:83)
! at com.newrelic.synthetics.minion.containersystemdriver.kubernetes.KubernetesHelper.getPodContainerLogs(KubernetesHelper.java:86)
! ... 22 common frames omitted

2020-11-25 00:00:59,140 - The node where job synthetic-healthcheck-93d8c8163bd145938020dc0de43aa1c2 was scheduled did not have enough available resources
2020-11-25 00:00:59,140 - Health check job synthetic-healthcheck-93d8c8163bd145938020dc0de43aa1c2 exited with 126
2020-11-25 00:00:59,140 - Runner Docker Image: synthetics-minion-runner:3.0.28 - Container System Driver (with KubernetesClient) failed completing the health check Job, as it returned with status failed.
2020-11-25 00:00:59,886 - Service/dependency 'ContainerSystemDriverHealthCheck' is not healthy 'Result{isHealthy=false, message=ContainerSystemDriver is not healthy: Runner Docker Image: synthetics-minion-runner:3.0.28 - Container System Driver (with KubernetesClient) failed completing the health check Job, as it returned with status failed., timestamp=2020-11-25T00:00:59.202Z}'
2020-11-25 00:00:59,887 -
2020-11-25 00:00:59,887 - **************************************************************************************************
2020-11-25 00:00:59,887 - ***                       MINION FAILED FIRST HEALTH CHECK (see above)                         ***
2020-11-25 00:00:59,887 - **************************************************************************************************
2020-11-25 00:00:59,887 -
2020-11-25 00:00:59,887 - *** QUITTING FORCEFULLY ***

      Exit Code:  1
      Started:    Tue, 24 Nov 2020 18:00:46 -0600
      Finished:   Tue, 24 Nov 2020 18:01:00 -0600
    Last State:   Terminated
      Reason:     Error
      Message:    .fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:568)
! at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:507)
! at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.doGetLog(PodOperationsImpl.java:186)
! at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.getLog(PodOperationsImpl.java:195)
! at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.getLog(PodOperationsImpl.java:83)
! at com.newrelic.synthetics.minion.containersystemdriver.kubernetes.KubernetesHelper.getPodContainerLogs(KubernetesHelper.java:86)
! ... 22 common frames omitted

2020-11-24 23:55:34,430 - The node where job synthetic-healthcheck-7702253108e349abac33e5d9ef95044a was scheduled did not have enough available resources
2020-11-24 23:55:34,430 - Health check job synthetic-healthcheck-7702253108e349abac33e5d9ef95044a exited with 126
2020-11-24 23:55:34,430 - Runner Docker Image: synthetics-minion-runner:3.0.28 - Container System Driver (with KubernetesClient) failed completing the health check Job, as it returned with status failed.
2020-11-24 23:55:35,412 - Service/dependency 'ContainerSystemDriverHealthCheck' is not healthy 'Result{isHealthy=false, message=ContainerSystemDriver is not healthy: Runner Docker Image: synthetics-minion-runner:3.0.28 - Container System Driver (with KubernetesClient) failed completing the health check Job, as it returned with status failed., timestamp=2020-11-24T23:55:34.444Z}'
2020-11-24 23:55:35,413 -
2020-11-24 23:55:35,413 - **************************************************************************************************
2020-11-24 23:55:35,413 - ***                       MINION FAILED FIRST HEALTH CHECK (see above)                         ***
2020-11-24 23:55:35,413 - **************************************************************************************************
2020-11-24 23:55:35,413 -
2020-11-24 23:55:35,413 - *** QUITTING FORCEFULLY ***

      Exit Code:    1
      Started:      Tue, 24 Nov 2020 17:55:22 -0600
      Finished:     Tue, 24 Nov 2020 17:55:35 -0600
    Ready:          False
    Restart Count:  12
    Limits:
      cpu:     750m
      memory:  1717986918400m
    Requests:
      cpu:     500m
      memory:  800Mi
    Liveness:  http-get http://:http/status/check delay=600s timeout=60s period=300s #success=1 #failure=3
    Environment:
      MINION_POD_NAME:              newrelic-synthetics-minion-2 (v1:metadata.name)
      MINION_POD_NAMESPACE:         monitoring (v1:metadata.namespace)
      MINION_PRIVATE_LOCATION_KEY:  NRSP-usXXXXXXX
      MINION_LOG_LEVEL:             INFO
      MINION_HEAVY_WORKERS:         2
      MINION_LIGHTWEIGHT_WORKERS:   50
    Mounts:
      /tmp from minion-volume (rw,path="newrelic-synthetics-minion/tmp")
      /var/run/secrets/kubernetes.io/serviceaccount from newrelic-synthetics-minion-token-xppxf (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  minion-volume:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  minion-volume-newrelic-synthetics-minion-2
    ReadOnly:   false
  newrelic-synthetics-minion-token-xppxf:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  newrelic-synthetics-minion-token-xppxf
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  beta.kubernetes.io/os=linux
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                  Age                    From                                          Message
  ----     ------                  ----                   ----                                          -------
  Warning  FailedScheduling        <unknown>              default-scheduler                             error while running "VolumeBinding" filter plugin for pod "newrelic-synthetics-minion-2": pod has unbound immediate PersistentVolumeClaims
  Warning  FailedScheduling        <unknown>              default-scheduler                             error while running "VolumeBinding" filter plugin for pod "newrelic-synthetics-minion-2": pod has unbound immediate PersistentVolumeClaims
  Normal   Scheduled               <unknown>              default-scheduler                             Successfully assigned monitoring/newrelic-synthetics-minion-2 to aks-defaultpool-33270332-vmss000015
  Normal   NotTriggerScaleUp       40m                    cluster-autoscaler                            pod didn't trigger scale-up (it wouldn't fit if a new node is added):
  Normal   SuccessfulAttachVolume  39m                    attachdetach-controller                       AttachVolume.Attach succeeded for volume "pvc-cd859e31-7afe-47d0-8f3c-e5c19e355bd8"
  Normal   Pulling                 39m                    kubelet, aks-defaultpool-33270332-vmss000015  Pulling image "quay.io/newrelic/synthetics-minion:3.0.28"
  Normal   Pulled                  39m                    kubelet, aks-defaultpool-33270332-vmss000015  Successfully pulled image "quay.io/newrelic/synthetics-minion:3.0.28"
  Normal   Created                 39m                    kubelet, aks-defaultpool-33270332-vmss000015  Created container update-mounted-subpath-permissions
  Normal   Started                 39m                    kubelet, aks-defaultpool-33270332-vmss000015  Started container update-mounted-subpath-permissions
  Normal   Pulled                  37m (x5 over 39m)      kubelet, aks-defaultpool-33270332-vmss000015  Container image "quay.io/newrelic/synthetics-minion:3.0.28" already present on machine
  Normal   Created                 37m (x5 over 39m)      kubelet, aks-defaultpool-33270332-vmss000015  Created container synthetics-minion
  Normal   Started                 37m (x5 over 39m)      kubelet, aks-defaultpool-33270332-vmss000015  Started container synthetics-minion
  Warning  BackOff                 4m31s (x150 over 38m)  kubelet, aks-defaultpool-33270332-vmss000015  Back-off restarting failed container
```
**What you expected to happen?**

It would not crash

**How to reproduce it?**

I would say spin up a node that is close the resource limits (but allows this to fit) and then add this to the node.
We're running in AKS 

**Anything else we need to know?**
Initially I doubled the cpu & memory resource that it requested, and it's limit.  That didn't correct the issue as it looked like it spun up a job that desired that amount of resources also so increasing it seemed to have it be harder to fit within a node.

Then
I set the replica set to 3 (thinking to maybe spread the load), it is running happily on the node (004) that is less utilized, but the 2 that are close to capacity it is crashlooping on.

```
 kubectl describe node aks-defaultpool-33270332-vmss000004 | grep Allocated -A12
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                       Requests          Limits
  --------                       --------          ------
  cpu                            2920m (75%)       9600m (248%)
  memory                         7342450944 (29%)  18432375910400m (73%)
  ephemeral-storage              0 (0%)            0 (0%)
  attachable-volumes-azure-disk  0                 0
Events:                          <none>
PS C:\Users\David.Bollman> kubectl describe node aks-defaultpool-33270332-vmss000012 | grep Allocated -A12
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                       Requests           Limits
  --------                       --------           ------
  cpu                            3431m (88%)        15850m (410%)
  memory                         10053780608 (39%)  27078187110400m (107%)
  ephemeral-storage              0 (0%)             0 (0%)
  attachable-volumes-azure-disk  0                  0
Events:                          <none>
PS C:\Users\David.Bollman> kubectl describe node aks-defaultpool-33270332-vmss000015 | grep Allocated -A7
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                       Requests           Limits
  --------                       --------           ------
  cpu                            3559m (92%)        23500m (608%)
  memory                         17081043584 (67%)  45003221606400m (178%)
  ephemeral-storage              0 (0%)             0 (0%)
  attachable-volumes-azure-disk  0                  0
```

**Comment by samb0803** [comment]
Same issue on AKS with 1cpu request / limit.

**Comment by samb0803** [comment]
Same issue on AKS.

My understanding is that the CPU request is coming from the job that is spun up through the code on the Newrelic Synthetic minions. The code creates a `job` to execute the synthetic task, and it sets the cpu request to `500mi`. 

But there's no way to interact with the job creation (through the Helm chart).

**Comment by keegoid-nr** [comment]
@samb0803 Is correct, the minion acts like a container orchestrator, spinning up runner pods and healthcheck pods as needed. Runner pods (and the resources they request) are not user configurable via the Helm chart.

You can set `heavyWorkers: "1"` instead of the default 2, which can help reduce the risk of running into this issue.
https://github.com/newrelic/helm-charts/blob/2a9949b330602f5bf16c3dabf0ca06a5bbc9c673/charts/synthetics-minion/values.yaml#L27

The true amount of resources required by the K8s CPM is not initially evident and hard to derive from the requirements doc.
https://docs.newrelic.com/docs/synthetics/synthetic-monitoring/private-locations/install-containerized-private-minions-cpms/#kubernetes-requirements

More details on required resources here:
https://newrelic.zendesk.com/knowledge/articles/4408123035287/en-us?brand_id=3270506

This issue is often encountered on clusters where the PVC is requesting access mode RWO since all pods will be scheduled on the same node. This makes setting a resource quota on the namespace more of a necessity to ensure other projects can't use resources requested by the CPM. This is especially true in a multi-tenant cluster.

**Comment by samb0803** [comment]
@keegoid-nr

That makes sense regarding the disk access mode. Essentially I should just need to change it to a storageclass that supports ReadWriteMany and specify the accessmode in the helm chart. 

I'm trying this now for AzureFile. I can see the RWX PVC has been bound by the pod. But the synthetic-minion errors with: 
`2021-09-23 01:47:41,082 - One and only one volume is expected to be bound to the minion synthetics-minion-0 - volumes found: [Volume(awsElasticBlockStore=null, azureDisk=null, azureFile=null, cephfs=null, cinder=null, configMap=null, csi=null, downwardAPI=null, emptyDir=null, fc=null, flexVolume=null, flocker=null, gcePersistentDisk=null, gitRepo=null, glusterfs=null, hostPath=null, iscsi=null, name=minion-volume, nfs=null, persistentVolumeClaim=PersistentVolumeClaimVolumeSource(claimName=minion-volume-synthetics-minion-0, readOnly=null, additionalProperties={}), photonPersistentDisk=null, portworxVolume=null, projected=null, quobyte=null, rbd=null, scaleIO=null, secret=null, storageos=null, vsphereVolume=null, additionalProperties={}), Volume(awsElasticBlockStore=null, azureDisk=null, azureFile=null, cephfs=null, cinder=null, configMap=null, csi=null, downwardAPI=null, emptyDir=null, fc=null, flexVolume=null, flocker=null, gcePersistentDisk=null, gitRepo=null, glusterfs=null, hostPath=null, iscsi=null, name=kube-api-access-pxvb7, nfs=null, persistentVolumeClaim=null, photonPersistentDisk=null, portworxVolume=null, projected=ProjectedVolumeSource(defaultMode=420, sources=[VolumeProjection(configMap=null, downwardAPI=null, secret=null, serviceAccountToken=ServiceAccountTokenProjection(audience=null, expirationSeconds=3607, path=token, additionalProperties={}), additionalProperties={}), VolumeProjection(configMap=ConfigMapProjection(items=[KeyToPath(key=ca.crt, mode=null, path=ca.crt, additionalProperties={})], name=kube-root-ca.crt, optional=null, additionalProperties={}), downwardAPI=null, secret=null, serviceAccountToken=null, additionalProperties={}), VolumeProjection(configMap=null, downwardAPI=DownwardAPIProjection(items=[DownwardAPIVolumeFile(fieldRef=ObjectFieldSelector(apiVersion=v1, fieldPath=metadata.namespace, additionalProperties={}), mode=null, path=namespace, resourceFieldRef=null, additionalProperties={})], additionalProperties={}), secret=null, serviceAccountToken=null, additionalProperties={})], additionalProperties={}), quobyte=null, rbd=null, scaleIO=null, secret=null, storageos=null, vsphereVolume=null, additionalProperties={})]`

Evidence of the pvc being bound and configured directly (storageClass + Access Mode). 
NAME                                STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
minion-volume-synthetics-minion-0   Bound    pvc-c55419e0-ce92-4fe5-93e4-28774865b23c   10Gi       RWX            azurefile      4m33s


Can you please confirm that ` PVC is requesting access mode RWO since all pods will be scheduled on the same node` is referring to the synthetic-minion > job's being bound to the same node. Not $replica's of the synthetic nodes being on the same node. 

Wouldn't it be better to leverage a different node pool with taints and selectors rather than using namespace quotas?

**Comment by samb0803** [comment]
Ignore my error. The cluster auto patched to 1.21.x which the docos shows won't work due to the SA token automount.

**Comment by davidgit** [comment]
Hi ,

The next best step is to visit New Relic Support, where you can engage the New Relic Support Community or open a support ticket depending on your support level. The support team is best positioned to assist with your specific needs.

Please provide a link to this GitHub issue when submitting your community post or support ticket.

Thanks!
